pallets/emergency-para-xcm/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ cumulus_pallet_parachain_system::Config<
+			CheckAssociatedRelayNumber = Pallet<Self>,
+			XcmpMessageHandler = Pallet<Self>,
+		> + pallet_message_queue::Config<QueuePausedQuery = Pallet<Self>>
+	{
+		/// Overarching event type
+		type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Used before check associated relay block number. It should be what would be passed to
+		/// `cumulus_pallet_parachain_system` if this pallet was not being used.
+		type CheckAssociatedRelayNumber: CheckAssociatedRelayNumber;
+
+		/// Used to check wether message queue is paused when `XcmMode` is `Normal`. It should be
+		/// what would be passed to `pallet_message_queue` if this pallet was not being used.
+		type QueuePausedQuery: QueuePausedQuery<<Self::MessageProcessor as ProcessMessage>::Origin>;
+
+		/// The XCMP handler to be used in normal operating mode
+		type XcmpMessageHandler: XcmpMessageHandler;
+
+		/// Maximum number of relay block to skip before trigering the Paused mode.
+		type PausedThreshold: Get<RelayChainBlockNumber>;
+
+		/// Origin allowed to perform a fast authorize upgrade when XcmMode is not normal
+		type FastAuthorizeUpgradeOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// origin allowed to resume to normal operations from paused mode
+		type PausedToNormalOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event {
+		/// The XCM incoming execution was Paused
+		EnteredPausedXcmMode,
+		/// The XCM incoming execution returned to normal operation
+		NormalXcmOperationResumed,
+	}
+
+	/// An error that can occur while executing this pallet's extrinsics.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The current XCM Mode is not Paused
+		NotInPausedMode,
+	}
+
+	#[pallet::storage]
+	/// Whether incoming XCM is enabled or paused
+	pub type Mode<T: Config> = StorageValue<_, XcmMode, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Account for 1 read and 1 write to `Mode`
+			T::DbWeight::get().reads_writes(1, 1)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Resume `Normal` mode
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 2), DispatchClass::Operational))]
+		pub fn paused_to_normal(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			T::PausedToNormalOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				Mode::<T>::get() == XcmMode::Paused,
+				Error::<T>::NotInPausedMode
+			);
+
+			Mode::<T>::set(XcmMode::Normal);
+			<Pallet<T>>::deposit_event(Event::NormalXcmOperationResumed);
+
+			Ok(().into())
+		}
+
+		/// Authorize a runtime upgrade. Only callable in `Paused` mode
+		#[pallet::call_index(1)]
+		#[pallet::weight((T::DbWeight::get().read, DispatchClass::Operational))]
```

Comment:
This extrinsic also writes in `authorize_upgrade`, we should account for it even if this is only executed by authorised accounts.